### ISSUE TITLE
Is init used when deploying via Captain?

### ISSUE THREAD

**Issue by drmrbrewer** [githubsaturn/captainduckduck#199]
I'm deploying an app with Captain that is based on [this dockerfile](https://github.com/skalfyfan/dockerized-puppeteer/blob/master/chrome/Dockerfile), and [this entrypoint](https://github.com/skalfyfan/dockerized-puppeteer/blob/master/chrome/entrypoint.sh).

The dockerfile is based on `krallin/ubuntu-tini:xenial`, which appears to be just a drop-in replacement for `ubuntu:xenial` but with a sort of [`dumb-init`](https://github.com/Yelp/dumb-init) functionality to prevent zombie processes... or rather an equivalent of `dumb-init` called [`tini`](https://github.com/krallin/tini).

In the README for `tini` it [states that](https://github.com/krallin/tini#using-tini):

> NOTE: If you are using Docker 1.13 or greater, Tini is included in Docker itself. This includes all versions of Docker CE. To enable Tini, just pass the --init flag to docker run.

Since, when using Captain to deploy the app, we never go near `docker run`, is this `--init` flag something that could be included by Captain, so that there is then no need to use `tini` as the base?  I did a search on `init` in the Captain code and couldn't see anything.

P.S. if you are able to give any hints on how I can scale this container up to multiple instances, this would be appreciated :)  The problem I encounter is explained [here](https://stackoverflow.com/q/49028742/4070848)... but so far nobody has suggested any answer.

**Comment by githubsaturn** [comment]
Captain uses Docker services it doesn't support `--init` flag yet. Docker people are supposed to copy all `docker run` functionalities over to `docker service`, but they are lagging. Anyways, for now, I'd suggest you use `tini` in your dockerfile.

PS:
About scalability of your app, I think nginx needs to be configured such that once you land on one instance, it keeps you on that instance for subsequent requests. For plain nginx, I would use `ip-hash` but for Docker+nginx, I don't think an out of the box solution can satisfy this need. Either change your architecture such that this is not a requirement, or use a custom load balancer, such as HA Proxy in front of a DNS round robin configured docker service using `--endpoint-mode dnsrr`

**Comment by drmrbrewer** [comment]
Regarding `init` not being available for a docker service, I guess the same seems to be the case for `--shm-size`?  Because another [app I'd like to deploy](https://github.com/joelgriffith/browserless#-docker-quickstart) needs to be run with `docker run --shm-size=1gb` but that option isn't available for `docker service create` :-(

[This issue](https://github.com/moby/moby/issues/26714) suggests that it's possible to use `--mount` as a workaround... how would that be set when deploying via Captain?

**Comment by githubsaturn** [comment]
Persistent directories on Captain are mounts.
- If you assign a map to host, they'll be the type of "bind"
- If you don't assign a map to host, and just assign a volume name, they'll be the type of "volume"

`tempfs` is not natively supported by Captain.

For your specific concern, you should set your application to have persistent data and also set the following mapping:
Container path: `/dev/shm`
Host path: `/dev/shm`

**Comment by drmrbrewer** [comment]
The issue then is that with persistent data enabled, the app is no longer scaleable (ref our other [other discussion](https://github.com/githubsaturn/captainduckduck/issues/198) :)

**Comment by githubsaturn** [comment]
Haha!

Okay so, don't set your app with persistent data, then manually update your service to use volumes. If the app is not set as a Persistent Data app, Captain does not modify the volumes. Therefore, you can manually update the app and assign volumes without worrying that Captain will override it. Steps:

- Create an app without persistent data
- SSH to server and run 
```
docker service update srv-captain--appname mount-add  \
                          type=bind,source=/dev/shm,target=/dev/shm
```

**Comment by drmrbrewer** [comment]
OK thanks.  That would need to be run how often... presumably not every time I deploy the app?

**Comment by githubsaturn** [comment]
Once - after app creation.

**Comment by drmrbrewer** [comment]
OK thanks.  I suppose, however, that I would potentially run into the sort of issues you described in the above linked issue... if multiple containers within the same service (and on the same node) are accessing the same volume then they could be trying to write to the same files at the same time...

**Comment by githubsaturn** [comment]
In theory yes - it'd be prone to the same problem. However, `/dev` directory on linux based systems are not your conventional directories. Linux file systems are quite complicated and powerful, in your example `/dev/shm` is just a chunk of memory that is being shared between different processes. On a properly design system, this path "is" supposed to be shared between different processes. So I don't think it'll cause issues.

**Comment by drmrbrewer** [comment]
OK thanks for your help :)

**Comment by drmrbrewer** [comment]
Just trying out mounting `/dev/shm`...

Firstly, there's a small typo in your above post... should be `--mount-add` rather than `mount-add`.

Secondly, should "Once - after app creation" be "Every time the app is redeployed or restarted, e.g. by hitting Save & Update Configuration"?

It seems, at least for me, that every time I restart the service/app, the `/dev/shm` mount is lost.  Is there any possibility that we could be given the ability to pass custom options to `docker service create`, or some other way of making this sort of thing more automated / fool-proof?

**Comment by githubsaturn** [comment]
> Secondly, should "Once - after app creation" be "Every time the app is redeployed or restarted, e.g. by hitting Save & Update Configuration"?

My bad. The service gets updated every time you deploy and update the app settings. 

> Is there any possibility that we could be given the ability to pass custom options to docker service create, or some other way of making this sort of thing more automated/fool-proof?

No - Captain runs itself in a container and does not have access to CLI docker commands, it talks to docker via docker unix socket API. 

> Or, allow the setting of the persistent directories info without also preventing app scaling... ;-) Couldn't app scaling be prevented by default, but with a checkbox that can be unchecked to allow scaling, and with a health warning by it?

It is possible. But it doesn't have any web dashboard support!
- Create an app WITHOUT persistent data.
- Go to apps settings and set all your environmental variables and other settings
- Save and Update Configurations
- `nano /captain/config.conf`
- Find your app and manually change the volume from an empty array `[]` to:
```
			"volumes": [
				{
					"containerPath": "/dev/shm",
					"hostPath": "/dev/shm",
					"type": "bind"
				}
			],
```

Done :heavy_check_mark: 

The only caveat is that once you do this, you won't be able to update your app via captain dashboard. It'll give you an error that you're setting volume for an app without persistent data settings. All your updates have to be done manually on `/captain/config.conf`

**Comment by drmrbrewer** [comment]
OK thanks, that's another possible workaround.  Not ideal, because I would need to be able to change an environment variable and restart this app, fairly frequently.

Regarding a more permanent solution, I have a straightforward suggestion...

You mention above that `tmpfs` is not natively supported by Captain.  Well, if it were, then this would completely solve my problem, and make everything customisable via Captain.  Furthermore, it would honour your pledge not to allow app scaling when persistent storage is enabled.  This is because with `tmpfs` there is no link back to a specific location on the host filesystem.  As I understand it, `tmpfs` is entirely self-contained, and lives and dies with the container.  It is not expected to outlive the container... the clue is in the name `tmpfs` (temporary filesystem).  Therefore it's a perfect fit for a scalable (inherently temporary) container, isn't it?

In particular, I find that running the following (on the host server) works nicely to set up a `tmpfs` folder inside the service's containers, mounted at `/dev/shm` to replace the tiny one that is there by default:

    # docker service update srv-captain--myapp --mount-add type=tmpfs,target=/dev/shm

In the resulting container:

    # df -h
    Filesystem      Size  Used Avail Use% Mounted on
    tmpfs           1.9G   67M  1.9G   4% /dev/shm

So, in the app settings on the Captain console, you would just have a simple new option for "tmpfs directories" (or something more plain-English), with just an editable `path` field and a button for adding more if needed:

    Temporary Directories
    Path in App:  ___________________
    [Add Directory]

You could also allow the size of each `tmpfs` mount to be specified, as an option (`tmpfs-size`).

What about it?

The docker documentation for `tmpfs` mounts can be found [here](https://docs.docker.com/storage/tmpfs/).

**Comment by githubsaturn** [comment]
You are correct. But this will only address one edge case. It will be too much effort for little gain. I need a better solution for other edge cases. I am thinking of customizing post-deploy and post-update hooks, i.e., in your case, you would simply define a post-deploy / post-save script with the following content:
```
docker service update srv-captain--myapp --mount-add type=tmpfs,target=/dev/shm
```
This will allow custom scripts to get executed after deploy and save. And will allow for all edge case addressing.

**Comment by drmrbrewer** [comment]
With the post deploy script, wouldn't the service go live briefly, and then immediately be taken down and restarted?

Re `tmpfs` it just seems to be begging to be added as an explicit deploy option... its siblings `volume` and `bind` are already represented and right now it is feeling like the ugly duckling of captain duck duck :). As for "too much effort", it seems that this would be one of the easier additions that could be made... almost copy and paste!

**Comment by githubsaturn** [comment]
Design note:
The mission of Captain is to provide users with a simple interface to perform "standard actions" - similar to Heroku. Mirroring Docker capabilities 1:1 is just very wrong in terms of product design. There's a point where the project gets so complicated that it demands a more complicated deployment system. Compared to Heroku, CaptainDuckDuck has a lot more features. However, it is still behind Docker in terms of complexity. There needs to be a point where we draw the line and say this is too edge case for a simple deployment. `tmpfs` is surely one of them. Adding tmpfs requires adding new radio buttons or lists, which makes things complicated for 99.9% of users who don't care about this and they get confused what to choose. On the other hand, post-deploy script opens up a new set of functionality. In terms of project going live, no, it will be immediately after deploy. So before your app starts up, the script gets executed.

**Comment by drmrbrewer** [comment]
So the post deploy script won't increase the time between stopping the old app and starting the new one?

**Comment by githubsaturn** [comment]
It does. Probably 1 sec or so.

**Comment by drmrbrewer** [comment]
When I run `docker service update` manually after deploying the app, to add the `tmpfs` mount, the time taken is more like 10 seconds... 5 seconds to do the actual update, and another 5 seconds waiting to ensure that the "tasks are stable".  So maybe it's only the first 5 seconds that count, but it's still an extra 5 seconds... maybe 10.  So, whilst a post-deploy script functionality would be very useful, it would be better to address this issues like these during the deploy, not after, if possible.

As for being an edge case, well I'm not so sure.  It seems that the miserly 64MB allocated to `/dev/shm` is a problem for many dockerised apps.  Look at the results of [this google search](https://www.google.com/search?q=docker+%2Fdev%2Fshm+64MB&oq=docker+%2Fdev%2Fshm+64MB) (`docker /dev/shm 64MB`)... lots of people having issues and needing to increase the size of `/dev/shm`.  Docker already provides a solution via e.g. `docker run --shm-size=2gb` but this isn't possible when deploying via Captain.

**Comment by githubsaturn** [comment]
I understand your points and concerns. But `/dev/shm` is not commonly used for web apps. If we go with the google search logic, we have to implement everything that docker offers, have a look at [this](https://www.google.ca/search?ei=QxWiWr77A9XejwPR3J6gCQ&q=docker+rollback+&oq=docker+rollback+&gs_l=psy-ab.3..0l4j0i22i30k1l2.12153.12153.0.12341.1.1.0.0.0.0.66.66.1.1.0....0...1c.1.64.psy-ab..0.1.65....0.WsOFGHr00Fw), and [this](https://www.google.ca/search?ei=lhWiWojGEY7qjwOFo4-ABw&q=docker+kernel+memory&oq=docker+kernel-memory&gs_l=psy-ab.3.0.0i30k1l2.1366.2236.0.3141.7.7.0.0.0.0.130.612.4j3.7.0....0...1c.1.64.psy-ab..3.4.387...0i7i30k1j0i13k1j0i8i7i30k1.0.kIHRGPdItfQ), and [this](https://www.google.ca/search?ei=mhWiWqi8FsWgjwP-lpGICg&q=docker+limit+cpu&oq=docker+limit+cpu&gs_l=psy-ab.3..0j0i22i30k1l6.15524.20080.0.20516.13.13.0.0.0.0.89.843.13.13.0....0...1c.1.64.psy-ab..0.13.837...35i39k1j0i67k1j0i20i263k1.0.3NJR8wswngo), for example. Various projects do have special needs, and that's where plain Docker with all bells and whistles comes into play.

When I called this case, an edge case, I made my comment considering the fact that this project is live and used by thousands of devs around the world, and after half a year, there is only one request for this. I am not married to "not changing anything". If the need is there, other people would come and say the same thing, then I put this issue on my radar and work on it.

Last but not least, CaptainDuckDuck is free, open source under MIT license - which means you can fork the project and do special things that you want to do if some changes help your process :)

**Comment by githubsaturn** [comment]
> When I run docker service update manually after deploying the app, to add the tmpfs mount, the time taken is more like 10 seconds... 5 seconds to do the actual update, and another 5 seconds waiting to ensure that the "tasks are stable". So maybe it's only the first 5 seconds that count, but it's still an extra 5 seconds... maybe 10. So, whilst a post-deploy script functionality would be very useful, it would be better to address this issues like these during the deploy, not after, if possible.

CaptainDuckDuck uses docker UNIX socket API. To see what it actually adds to your deployment time, try adding `--detach` flag to `docker update` command.

**Comment by drmrbrewer** [comment]
> after half a year, there is only one request for this

but I've made this request *at least three times* in this thread ;-)

Seriously, I understand completely, and I admire your restraint.  If this were my project, by now it would have a thousand little adhoc add-ons and would be completely unusable by anyone but myself.

BTW this issue arises (amongst others) when running chrome inside a docker container, since chrome appears by default to make use of `/dev/shm`, and for many practical uses the default 64MB is completely insufficient.  The [puppeteer](https://github.com/GoogleChrome/puppeteer) project provides an API for controlling chrome in such a context, and with nearly 30,000 stars is pretty popular (even more popular than your project ;-).  However, they have in fact now provided an explicit option to pass to puppeteer, independent of docker options, to disable usage of `/dev/shm` entirely, so that's what I'll use.  It would be much better IMHO to retain usage of `/dev/shm` (the chrome folks must be using it for good reason) and just increase its size, but it seems that this isn't (practically) possible for now with Captain.

**Comment by githubsaturn** [comment]
Yes - I know of puppeteer - big fan of it! I totally agree with you that 64MB is pretty much useless. I'll do more research on this, and based on my findings, I'll do one of these:
- Default to host `/dev/shm` and have the ability for users to opt out through a config file - not part of the default GUI.
- Leave as is and have the ability for users to opt out through a config file - not part of the default GUI.
- Go with a generic post deploy solution
- Go with a pre-processor of `service update` such that `service update` gets called only once, but user can add/remove params.

**Comment by githubsaturn** [comment]
Predeploy script added in https://github.com/githubsaturn/captainduckduck/commit/d56c31909ab2b5c8e8265500f99db6af446ec9c7

WIll be released in v0.7.0

**Comment by drmrbrewer** [comment]
> Captain uses Docker services it doesn't support `--init` flag yet. 

It seems like the `--init` flag is now supported?  See [current option list](https://docs.docker.com/engine/reference/commandline/service_create/#options).  How do we specify `--init` with CapRover?

**Comment by githubsaturn** [comment]
It was added in API 1.37 - CapRover uses 1.30 right now. Technically the docker API version should have been part of the [customizable configs](https://caprover.com/docs/troubleshooting.html#customize-config-settings) but it was missed.

I've create an issue to address that. After this work was done, you can manually bump up your API version and you can add `init` like this:
```
var preDeployFunction = function (captainAppObj, dockerUpdateObject) {
    return Promise.resolve()
        .then(function(){

            dockerUpdateObject.TaskTemplate.ContainerSpec.init = true;
            return dockerUpdateObject;
        });
};
```

**Comment by drmrbrewer** [comment]
OK thanks.  Rather than having to update the pre-deploy script in each and every app in the console (and I have many), my feeling is that it may just be easier to leave things as they are... i.e. installing init/tini and using it as the entry point in the Dockerfile.

**Comment by hroland** [comment]
Related to this thread, for anyone wondering how to make volume mounts work with pre-deploy scripts:

```
var preDeployFunction = function (captainAppObj, dockerUpdateObject) {
    return Promise.resolve()
        .then(function(){

            dockerUpdateObject.TaskTemplate.ContainerSpec.Mounts = [
                { Type: 'bind', Source: '/var/run/docker.sock', Target: '/var/run/docker.sock' }
            ];
            return dockerUpdateObject;
        });
};
```
